### PR TITLE
AlternativeTextType as enum class with EnumTraits and serialization

### PR DESCRIPTION
**Pull request by angelaizg**
#### d84ac58bd76d15e40b36f215e7c8ecae110686d2
<pre>
AlternativeTextType as enum class with EnumTraits and serialization
<a href="https://bugs.webkit.org/show_bug.cgi?id=248174">https://bugs.webkit.org/show_bug.cgi?id=248174</a>
rdar://102309622

Making AlternativeTextType an enum class with EnumTraits and serialize it instead of casting to/from an int32_t. Prevents from assertion failure in void _NSWindowSetFrameIvar

Reviewed by Youenn Fablet

Canonical link: <a href="https://commits.webkit.org/257360@main">https://commits.webkit.org/257360@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/ebad6030560c521efafc3e0361b384eae04b48a6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98287 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7495 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31425 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107737 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168007 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102230 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7999 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85015 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90862 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/104323 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103941 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6025 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89584 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33101 "Passed tests") | 
| [❌ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/101833 "webkitpy-tests (failure)") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87895 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21001 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76045 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1454 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22521 "Found 1 new test failure: streams/readable-stream-default-controller-error.html (failure)") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1398 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45136 "Found 30 new test failures: accessibility/aria-keyshortcuts.html, css3/supports-dom-api.html, fast/css/css-typed-om/style-property-map-set-CSSMathSum-value.html, fast/rendering/render-style-null-optgroup-crash.html, fast/text/text-edge-no-half-leading-simple.html, fast/text/text-edge-property-parsing.html, http/wpt/service-workers/fetch-service-worker-preload-changing-request.https.html, http/wpt/service-workers/fetch-service-worker-preload-download.https.html, http/wpt/service-workers/fetch-service-worker-preload.https.html, imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml.html ... (failure)") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5052 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6292 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41942 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2748 "Built successfully") | | |

### CODE REVIEW

File: LayoutTests/ipc/WebPageProxy_CorrectionPanel-expected.txt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+CONSOLE MESSAGE: ReferenceError: Can't find variable: IPC
```
Comment (rniwa) [inline LayoutTests/ipc/WebPageProxy_CorrectionPanel-expected.txt]: AH... looks like this test is broken?

File: Source/WebCore/loader/icon/IconLoader.cpp [hunk 2]
```diff
@@ -56,6 +56,7 @@ IconLoader::~IconLoader()
 
 void IconLoader::startLoading()
 {
+    WTFLogAlways("*************IconLoader::startLoading******************");
```
Comment (rniwa) [inline Source/WebCore/loader/icon/IconLoader.cpp]: This seems like an unrelated change.

File: Source/WebCore/loader/cache/CachedResourceLoader.cpp [hunk 3]
```diff
@@ -336,7 +336,6 @@ ResourceErrorOr<CachedResourceHandle<CachedRawResource>> CachedResourceLoader::r
     default:
         ASSERT_NOT_REACHED();
     }
-
```
Comment (rniwa) [inline Source/WebCore/loader/cache/CachedResourceLoader.cpp]: These whitespace changes seem untreated. Revert?

File: Source/WebCore/platform/graphics/WebMResourceClient.cpp [hunk 4]
```diff
@@ -36,6 +36,7 @@ namespace WebCore {
 
 RefPtr<WebMResourceClient> WebMResourceClient::create(WebMResourceClientParent& parent, PlatformMediaResourceLoader& loader, ResourceRequest&& request)
 {
+    WTFLogAlways("*************WebMResourceClient::create********");
```
Comment (rniwa) [inline Source/WebCore/platform/graphics/WebMResourceClient.cpp]: This seems like unrelated change.

**Review by rniwa — CHANGES_REQUESTED** [review]

File: LayoutTests/ipc/WebPageProxy_CorrectionPanel-expected.txt (hunk 1 above)
Comment (angelaizg) [reply LayoutTests/ipc/WebPageProxy_CorrectionPanel-expected.txt]: It is supposed to happen, since the int sent is not in the enum cases

File: Source/WebKit/UIProcess/WebPageProxy.cpp [hunk 5]
```diff
@@ -9672,10 +9674,14 @@ void WebPageProxy::substitutionsPanelIsShowing(CompletionHandler<void(bool)>&& c
     completionHandler(TextChecker::substitutionsPanelIsShowing());
 }
 
-void WebPageProxy::showCorrectionPanel(int32_t panelType, const FloatRect& boundingBoxOfReplacedString, const String& replacedString, const String& replacementString, const Vector<String>& alternativeReplacementStrings)
+void WebPageProxy::showCorrectionPanel(AlternativeTextType panelType, const FloatRect& boundingBoxOfReplacedString, const String& replacedString, const String& replacementString, const Vector<String>& alternativeReplacementStrings)
 {
     // FIXME: Make AlternativeTextType an enum class with EnumTraits and serialize it instead of casting to/from an int32_t.
-    pageClient().showCorrectionPanel((AlternativeTextType)panelType, boundingBoxOfReplacedString, replacedString, replacementString, alternativeReplacementStrings);
+    if (panelType == AlternativeTextType::AlternativeTextTypeDictationAlternatives) {
+        ASSERT_IS_TESTING_IPC();
```
Comment (youennf) [inline Source/WebKit/UIProcess/WebPageProxy.cpp]: It would be good to add a test for this code path.

File: Source/WebKit/UIProcess/WebPageProxy.cpp [hunk 6]
```diff
@@ -858,11 +859,12 @@ void WebPageProxy::setResourceLoadClient(std::unique_ptr<API::ResourceLoadClient
 
 void WebPageProxy::handleMessage(IPC::Connection& connection, const String& messageName, const WebKit::UserData& messageBody)
 {
+    
     ASSERT(m_process->connection() == &connection);
 
     if (!m_injectedBundleClient)
         return;
-
+
```
Comment (youennf) [inline Source/WebKit/UIProcess/WebPageProxy.cpp]: These two changes are not needed.

File: Source/WebKitLegacy/mac/WebCoreSupport/CorrectionPanel.mm [hunk 7]
```diff
@@ -35,18 +35,18 @@
...
-    case AlternativeTextTypeDictationAlternatives:
+    case AlternativeTextType::AlternativeTextTypeDictationAlternatives:
         ASSERT_NOT_REACHED();
-        break;
+        return NSCorrectionIndicatorTypeDefault;
+
     }
-    ASSERT_NOT_REACHED();
-    return NSCorrectionIndicatorTypeDefault;
+
```
Comment (youennf) [inline Source/WebKitLegacy/mac/WebCoreSupport/CorrectionPanel.mm:48]: It seems we could probably share correctionIndicatorType with CorrectionPanel to reduce redundant code.

File: LayoutTests/ipc/WebPageProxy_CorrectionPanel.html [hunk 8]
```diff
@@ -0,0 +1,4 @@
+<script>
```
Comment (youennf) [inline LayoutTests/ipc/WebPageProxy_CorrectionPanel.html]: This is probably not right, other IPC tests have:
`<!DOCTYPE html><!-- webkit-test-runner [ IPCTestingAPIEnabled=true ] -->`

This might explain the issue @rniwa spotted.

**Review by youennf — CHANGES_REQUESTED** [review]

**Comment by youennf** [comment]
You also have 3 commits, could you squash them and update this PR description accordingly?

File: Source/WebKit/UIProcess/WebPageProxy.cpp (hunk 5 above)
Comment (kkinnunen-apple) [reply Source/WebKit/UIProcess/WebPageProxy.cpp]: Change this to 
`MESSAGE_CHECK(panelType != AlternativeTextType::AlternativeTextTypeDictationAlternatives);`

It's unclear to me if the test can be added to the layout test after you do this. You can check and see if it crashes or is skipped.

File: LayoutTests/ipc/WebPageProxy_CorrectionPanel.html (hunk 8 above)
Comment (kkinnunen-apple) [reply LayoutTests/ipc/WebPageProxy_CorrectionPanel.html]: If the test case is possible to be added, it could be named something like webpageproxy-correctionpanel-no-crash.html. See the other files in the directory for examples how to write the test.

Comment (kkinnunen-apple) [reply LayoutTests/ipc/WebPageProxy_CorrectionPanel.html]: Also:
Originally the value was size of int32_t and the problem was that valid values were 0-3 but 655 was sent.

Now your code changes the value to size of uint8_t. 
So now the test just tests that the IPC wire format checks that the message is correct sized. The test will fail because int32_t is sent instead of uint8_t.

This means that the test case will not check the original underlying bug, e.g. sending values outside the allowed enum range.

So what you could do is to send couple of messages with different values > 3, but with uint8_t type. So this way you could ensure that the original bug is tested, e.g. that the function only allows the values in the enum.

Also, you could then add the case for the disallowed enum value, e.g.  `AlternativeTextTypeDictationAlternatives`.  This is with the caveat that it's not entirely clear to me if `MESSAGE_CHECK`ed values can be tested, or will that too crash.

File: LayoutTests/ipc/webpageproxy-correctionpanel-no-crash-expected.txt [hunk 9]
```diff
@@ -0,0 +1 @@
+
```
Comment (youennf) [inline LayoutTests/ipc/webpageproxy-correctionpanel-no-crash-expected.txt]: I would tend to do like the other tests and add something like
`Test passes if it does not crash.` in the html and which would show up there.

File: Source/WebKit/UIProcess/mac/CorrectionPanel.mm [hunk 10]
```diff
@@ -35,13 +35,13 @@
...
+    case WebCore::AlternativeTextType::AlternativeTextTypeCorrection:
         return NSCorrectionIndicatorTypeDefault;
-    case WebCore::AlternativeTextTypeReversion:
+    case WebCore::AlternativeTextType::AlternativeTextTypeReversion:
         return NSCorrectionIndicatorTypeReversion;
-    case WebCore::AlternativeTextTypeSpellingSuggestions:
+    case WebCore::AlternativeTextType::AlternativeTextTypeSpellingSuggestions:
         return NSCorrectionIndicatorTypeGuesses;
-    case WebCore::AlternativeTextTypeDictationAlternatives:
+    case WebCore::AlternativeTextType::AlternativeTextTypeDictationAlternatives:
```
Comment (youennf) [inline Source/WebKit/UIProcess/mac/CorrectionPanel.mm]: This one is now different from the WebKitLegacy one, let's keep them the same

**Comment by youennf** [comment]
The commit message is wrong, could you fix it so that it has the correct formatting.
You can also use git pr to resync the PR description with the commit message.

**Review by youennf — COMMENTED** [review]
Does it make sense to add a special test for AlternativeTextType::AlternativeTextTypeDictationAlternatives?

File: LayoutTests/ipc/webpageproxy-correctionpanel-no-crash-expected.txt (hunk 9 above)
Comment (angelaizg) [reply LayoutTests/ipc/webpageproxy-correctionpanel-no-crash-expected.txt]: I just added it in the parameters of that test

**Comment by angelaizg** [comment]
Failure of tests not related

File: Source/WebCore/page/AlternativeTextClient.h [hunk 11]
```diff
@@ -38,11 +38,11 @@ enum ReasonForDismissingAlternativeText {
     ReasonForDismissingAlternativeTextAccepted
 };
 
-enum AlternativeTextType {
+enum class AlternativeTextType : uint8_t {
     AlternativeTextTypeCorrection = 0,
     AlternativeTextTypeReversion,
     AlternativeTextTypeSpellingSuggestions,
-    AlternativeTextTypeDictationAlternatives
+    DictationAlternatives
```
Comment (youennf) [inline Source/WebCore/page/AlternativeTextClient.h:45]: It seems a good idea to rename AlternativeTextTypeDictationAlternatives to DictationAlternatives.
Can we do the same for the others 3 as well now?

File: Source/WebKit/UIProcess/mac/CorrectionPanel.mm [hunk 12]
```diff
@@ -35,18 +35,17 @@
...
         return NSCorrectionIndicatorTypeReversion;
-    case WebCore::AlternativeTextTypeSpellingSuggestions:
+    case WebCore::AlternativeTextType::AlternativeTextTypeSpellingSuggestions:
         return NSCorrectionIndicatorTypeGuesses;
-    case WebCore::AlternativeTextTypeDictationAlternatives:
+    case WebCore::AlternativeTextType::DictationAlternatives:
         ASSERT_NOT_REACHED();
-        break;
+        return NSCorrectionIndicatorTypeDefault;
+
```
Comment (youennf) [inline Source/WebKit/UIProcess/mac/CorrectionPanel.mm]: Unneeded line

File: Source/WebKitLegacy/mac/WebCoreSupport/CorrectionPanel.mm [hunk 13]
```diff
@@ -35,18 +35,18 @@
...
         return NSCorrectionIndicatorTypeReversion;
-    case AlternativeTextTypeSpellingSuggestions:
+    case AlternativeTextType::AlternativeTextTypeSpellingSuggestions:
         return NSCorrectionIndicatorTypeGuesses;
-    case AlternativeTextTypeDictationAlternatives:
+    case AlternativeTextType::DictationAlternatives:
         ASSERT_NOT_REACHED();
-        break;
+        return NSCorrectionIndicatorTypeDefault;
+
```
Comment (youennf) [inline Source/WebKitLegacy/mac/WebCoreSupport/CorrectionPanel.mm]: Unneeded line

File: Source/WebCore/page/AlternativeTextClient.h [hunk 14]
```diff
@@ -68,3 +68,17 @@ class AlternativeTextClient {
...
+namespace WTF {
+
+template<> struct EnumTraits<WebCore::AlternativeTextType> {
+    using values = EnumValues<
+        WebCore::AlternativeTextType,
+        WebCore::AlternativeTextType::AlternativeTextTypeCorrection,
+        WebCore::AlternativeTextType::AlternativeTextTypeReversion,
+        WebCore::AlternativeTextType::AlternativeTextTypeSpellingSuggestions,
+        WebCore::AlternativeTextType::DictationAlternatives
+
```
Comment (youennf) [inline Source/WebCore/page/AlternativeTextClient.h]: Unneeded line

File: LayoutTests/ipc/webpageproxy-correctionpanel-no-crash.html [hunk 15]
```diff
@@ -0,0 +1,14 @@
+<!DOCTYPE html><!-- webkit-test-runner [ IPCTestingAPIEnabled=true ] -->
+<html>
+<body>
+<script>
+
+if (window.testRunner)
+    testRunner.dumpAsText();
+
+if (window.IPC)
+    IPC.sendMessage('UI',IPC.webPageProxyID,IPC.messages.WebPageProxy_ShowCorrectionPanel.name,[{type: 'int32_t',value: 120},{type: 'float',value: 134.3630587066791},{type: 'float',value: 659.5882936421311},{type: 'float',value: 249},{type: 'uint8_t',value: 3},{type: 'String',value: ""},{type: 'String',value: ""},{type: 'Vector',value: [[{type: 'String',value: "cancelOperation"}],[{type: 'String',value: ""}],[{type: 'String',value: "page-9"}],[{type: 'String',value: ""}],[{type: 'String',value: "cancelOperation"}],[{type: 'String',value: "file:///tmp/ipcfuzz"}],[{type: 'String',value: ""}],[{type: 'String',value: "com.apple.pasteboard.clipboard"}]]}]);
```
Comment (youennf) [inline LayoutTests/ipc/webpageproxy-correctionpanel-no-crash.html]: Shouldn't we pass a uint8_t whose value is 120 now that we are serialising a uint8_t enum?

**Review by youennf — CHANGES_REQUESTED** [review]

File: LayoutTests/ipc/webpageproxy-correctionpanel-no-crash.html (hunk 15 above)
Comment (angelaizg) [reply LayoutTests/ipc/webpageproxy-correctionpanel-no-crash.html]: why is that?

File: Source/WebCore/page/AlternativeTextClient.h (hunk 11 above)
Comment (angelaizg) [reply Source/WebCore/page/AlternativeTextClient.h:45]: Added changes

File: Source/WebKit/UIProcess/WebPageProxy.cpp [hunk 16]
```diff
@@ -9680,10 +9681,10 @@ void WebPageProxy::substitutionsPanelIsShowing(CompletionHandler<void(bool)>&& c
     completionHandler(TextChecker::substitutionsPanelIsShowing());
 }
 
-void WebPageProxy::showCorrectionPanel(int32_t panelType, const FloatRect& boundingBoxOfReplacedString, const String& replacedString, const String& replacementString, const Vector<String>& alternativeReplacementStrings)
+void WebPageProxy::showCorrectionPanel(AlternativeTextType panelType, const FloatRect& boundingBoxOfReplacedString, const String& replacedString, const String& replacementString, const Vector<String>& alternativeReplacementStrings)
 {
     // FIXME: Make AlternativeTextType an enum class with EnumTraits and serialize it instead of casting to/from an int32_t.
-    pageClient().showCorrectionPanel((AlternativeTextType)panelType, boundingBoxOfReplacedString, replacedString, replacementString, alternativeReplacementStrings);
+    pageClient().showCorrectionPanel(panelType, boundingBoxOfReplacedString, replacedString, replacementString, alternativeReplacementStrings);
```
Comment (youennf) [inline Source/WebKit/UIProcess/WebPageProxy.cpp:9689]: Please remove above FIXME.

File: Source/WebCore/page/AlternativeTextClient.h (hunk 11 above)
Comment (youennf) [reply Source/WebCore/page/AlternativeTextClient.h:45]: Looks good now

**Review by youennf — APPROVED** [review]
r=me.
Please remove the no longer necessary FIXME.

File: Source/WebKit/UIProcess/WebPageProxy.cpp (hunk 16 above)
Comment (angelaizg) [reply Source/WebKit/UIProcess/WebPageProxy.cpp:9689]: Added changes

**Review by youennf — APPROVED** [review]

File: Source/WebKit/UIProcess/WebPageProxy.cpp (hunk 5 above)
Comment (angelaizg) [reply Source/WebKit/UIProcess/WebPageProxy.cpp]: Changes applied

File: Source/WebKitLegacy/mac/WebCoreSupport/CorrectionPanel.mm (hunk 7 above)
Comment (angelaizg) [reply Source/WebKitLegacy/mac/WebCoreSupport/CorrectionPanel.mm:48]: Opened a new bug a new patch that would take care of this

File: Source/WebCore/loader/cache/CachedResourceLoader.cpp (hunk 3 above)
Comment (angelaizg) [reply Source/WebCore/loader/cache/CachedResourceLoader.cpp]: Fixed

**Review by angelaizg — COMMENTED** [review]
Added the changes needed.
